mypyc/test-data/irbuild-match.test

```diff
@@ -0,0 +1,1678 @@
+    r1 = r0 != 0
+    if r1 goto L1 else goto L6 :: bool
+L1:
+    r2 = PyObject_Size(x)
+    r3 = r2 >= 0 :: signed
+    r4 = r2 >= 2 :: signed
+    if r4 goto L2 else goto L6 :: bool
+L2:
+    r5 = r2 - 2
+    r6 = PySequence_GetItem(x, r5)
+    r7 = object 1
+    r8 = PyObject_RichCompare(r6, r7, 2)
+    r9 = PyObject_IsTrue(r8)
+    r10 = r9 >= 0 :: signed
+    r11 = truncate r9: int32 to builtins.bool
+    if r11 goto L3 else goto L6 :: bool
+L3:
+    r12 = r2 - 1
+    r13 = PySequence_GetItem(x, r12)
+    r14 = object 2
+    r15 = PyObject_RichCompare(r13, r14, 2)
+    r16 = PyObject_IsTrue(r15)
+    r17 = r16 >= 0 :: signed
+    r18 = truncate r16: int32 to builtins.bool
+    if r18 goto L4 else goto L6 :: bool
+L4:
+    r19 = r2 - 2
+    r20 = PySequence_GetSlice(x, 0, r19)
+    rest = r20
+L5:
+    r21 = 'matched'
+    r22 = builtins :: module
+    r23 = 'print'
+    r24 = CPyObject_GetAttr(r22, r23)
+    r25 = [r21]
+    r26 = load_address r25
+    r27 = _PyObject_Vectorcall(r24, r26, 1, 0)
+    keep_alive r21
+    goto L7
+L6:
+L7:
+    r28 = box(None, 1)
+    return r28
+[case testMatchBuiltinClassPattern_python3_10]
+def f(x):
+    match x:
+        case int(y):
+            print("matched")
+[out]
+def f(x):
+    x, r0 :: object
+    r1 :: int32
+    r2 :: bit
+    r3 :: bool
+    r4, y :: int
+    r5 :: str
+    r6 :: object
+    r7 :: str
+    r8 :: object
+    r9 :: object[1]
+    r10 :: object_ptr
+    r11, r12 :: object
+L0:
+    r0 = load_address PyLong_Type
+    r1 = PyObject_IsInstance(x, r0)
+    r2 = r1 >= 0 :: signed
+    r3 = truncate r1: int32 to builtins.bool
+    if r3 goto L1 else goto L3 :: bool
+L1:
+    r4 = unbox(int, x)
+    y = r4
+L2:
+    r5 = 'matched'
+    r6 = builtins :: module
+    r7 = 'print'
+    r8 = CPyObject_GetAttr(r6, r7)
+    r9 = [r5]
+    r10 = load_address r9
+    r11 = _PyObject_Vectorcall(r8, r10, 1, 0)
+    keep_alive r5
+    goto L4
+L3:
+L4:
+    r12 = box(None, 1)
+    return r12
+[case testMatchSequenceCaptureAll_python3_10]
+def f(x):
+    match x:
+        case [*rest]:
+            print("matched")
+[out]
+def f(x):
+    x :: object
+    r0 :: int32
+    r1 :: bit
+    r2 :: native_int
+    r3, r4 :: bit
+    r5 :: native_int
+    r6, rest :: object
+    r7 :: str
+    r8 :: object
+    r9 :: str
+    r10 :: object
+    r11 :: object[1]
+    r12 :: object_ptr
+    r13, r14 :: object
+L0:
+    r0 = CPySequence_Check(x)
+    r1 = r0 != 0
+    if r1 goto L1 else goto L4 :: bool
+L1:
+    r2 = PyObject_Size(x)
+    r3 = r2 >= 0 :: signed
+    r4 = r2 >= 0 :: signed
+    if r4 goto L2 else goto L4 :: bool
+L2:
+    r5 = r2 - 0
+    r6 = PySequence_GetSlice(x, 0, r5)
+    rest = r6
+L3:
+    r7 = 'matched'
+    r8 = builtins :: module
+    r9 = 'print'
+    r10 = CPyObject_GetAttr(r8, r9)
+    r11 = [r7]
+    r12 = load_address r11
+    r13 = _PyObject_Vectorcall(r10, r12, 1, 0)
+    keep_alive r7
+    goto L5
+L4:
+L5:
+    r14 = box(None, 1)
+    return r14
```

Comment:
Can you add an irbuild test case that uses type annotations and native classes, to see whether we generate efficient code for these? E.g. something like this:

```py
class A:
    a: int

def f(x: A | int) -> int:
    match x:
        case A(a=a):
            return a
        case int():
            return x
```